task/jefe/src/dump.rs

```diff
@@ -119,56 +203,109 @@ pub fn dump_task(base: u32, task: usize) {
 
                 if let Err(e) = humpty::add_dump_segment_header(
-                    base,
+                    area.address,
                     region.base,
                     region.size,
                     |addr, buf, _| unsafe { humpty::from_mem(addr, buf) },
                     |addr, buf| unsafe { humpty::to_mem(addr, buf) },
                 ) {
                     ringbuf_entry!(Trace::DumpRegionsFailed(e));
-                    return;
+                    return Err(DumpAgentError::BadSegmentAdd);
                 }
 
                 ndx += 1;
             }
         }
     }
 
-    ringbuf_entry!(Trace::DumpStart(base));
+    dump_task_run(area.address, task)?;
 
-    //
-    // The humpty dance is your chance... to do the dump!
-    //
-    let r = humpty::dump::<(), 512, { humpty::DUMPER_JEFE }>(
+    // Convert from a dump area (address) back to an index
+    humpty::dump_address_to_index(base, area.address, |addr, buf, _| unsafe {
+        humpty::from_mem(addr, buf)
+    })
+    .map_err(|_| DumpAgentError::DumpFailed)
+}
+
+/// Dumps a specific region from the given task
+pub fn dump_task_region(
+    base: u32,
+    task: usize,
+    start: u32,
+    length: u32,
+) -> Result<u8, DumpAgentError> {
+    ringbuf_entry!(Trace::DumpingTaskRegion {
+        task,
         base,
-        Some(humpty::DumpTask::new(task as u16, sys_get_timer().now)),
-        || Ok(None),
-        |addr, buf, meta| {
-            ringbuf_entry!(Trace::DumpReading(addr, buf.len(), meta));
+        start,
+        length
+    });
 
-            //
-            // If meta is set, this read is metadata from within the dump
-            // regions (e.g., a dump header or dump segment header), and we'll
-            // read it directly -- otherwise, we'll ask the kernel.
-            //
-            if meta {
-                unsafe { humpty::from_mem(addr, buf) }
-            } else {
-                let r = kipc::read_task_dump_region(
-                    task,
-                    TaskDumpRegion {
-                        base: addr,
-                        size: buf.len() as u32,
-                    },
-                    buf,
-                );
+    if start & 0b11 != 0 {
+        return Err(DumpAgentError::UnalignedSegmentAddress.into());
+    }
 
-                ringbuf_entry!(Trace::DumpRead(r));
-                Ok(())
+    if (length as usize) & 0b11 != 0 {
+        return Err(DumpAgentError::UnalignedSegmentLength.into());
+    }
+
+    let area = dump_task_setup(base, false)?;
+
+    // We don't trust the caller; it may request to dump a region that isn't
+    // owned by this particular task!  To check this, we iterate over all of the
+    // valid dump regions and confirm that our desired region is within one of
+    // them.
+    let mut ndx = 0;
+    let mem = start..start + length;
+    let mut okay = false;
+    loop {
+        // This is Accidentally Quadratic; see the note in `dump_task`
```

Comment:
Either having the kernel return an error for an out-of-range region (as we discussed in chat), or exposing a "validate region" entry point, would be good ways of eliminating this

Reply:
Agreed, that's going on my list for later cleanups